oolab/src/main/java/agh/ics/oop/SimulationEngine.java

```diff
@@ -15,14 +22,41 @@ public SimulationEngine(MoveDirection[] directions, IWorldMap map, Vector2d[] po
         }
     }
 
-    public void run(){
-        int i = 0;
-        while (i < directions.length) {
-            for (Animal a : ((AbstractWorldMap) map).getAnimals()) {
-                if (i == directions.length) break;
-                a.move(directions[i]);
-                i++;
+    public SimulationEngine(AbstractWorldMap map_animals, Vector2d[] positions,App renderer){
+        this.renderer = renderer;
+        this.map = map_animals;
+        for (Vector2d position : positions) {
+            Animal animal = new Animal(this.map, position);
+            animal.addObserver(map_animals);
+            if (renderer != null){
+                animal.addObserver(renderer);
```

Comment:
miałoby to więcej sensu gdyby to `App` dodawało się jako obserwator, wtedy rzeczywiście `SimulationEngine` nie musi nic wiedzieć o istnieniu `App`. 

Swoja drogą, rozwiązanie z użyciem obserwatora na poszczególnych zwierzątkach (zamiast nowego jak w poleceniu) to fajny pomysł, tylko że tracimy wtedy informacje o zmianie orientacji - jak zwierzątko się kręci to na mapie tego nie widać, dopiero gdy się ruszy. Trzeba by zmodyfikować ten interfejs tak by notyfikował również o orientacjach.